### DISCUSSION TITLE
Integrating web3auth with PocketBase

### DISCUSSION THREAD

**Discussion by vpavlin** [pocketbase/pocketbase#2006] [Q&A] [answered]
Hi all, I've been digging into [web3auth](web3auth.io) for some time and trying to fully grasp its capabilities. In a nutshell, it allows you to login with various providers (social, password-less, external blockchain wallets) with relatively unified SDK and flow.

That said, it does not seem like they provide OAuth provider (no mention of OIDC in their doc), so it cannot be directly integrated into Pocketbase.

I am trying to figure out if there is a way around it - let's say I am fine trusting their JWT tokens (after signature verification using their JWKs) - is there a way  (e.g. custom middleware) to use web3auth JWT token, verify it and then authenticate PB user based on email or wallet address? 

Web3Auth JWT token has `email` and `wallets[0].public_key` fields in it (in case of social logins) or `wallets[0].address` field which is a blockchian wallet address. That means both typ

The flow I thought about was:
1. login with web3auth in frontend
2. Use the token in `pb.authStore.save(TOKEN, null)`
3. do `authRefresh`
4. Middleware intercepts, verifies the JWT token
5. Checks if user with wallet address from JWT exists
  a.  if not, create new user with email and/or wallet address
  b. if it does, load the user record
6. Proceed as usual authenticated request

Would love any feedback/comments whether this makes any sense and/or pointers to similar things being done as I am really new to Pocketbase, so still wrapping my head around it

**Comment by ganigeorgiev** [comment, marked as answer]
> is there a way (e.g. custom middleware) to use web3auth JWT token, verify it and then authenticate PB user based on email or wallet address?

Yes, PocketBase consider any user (or admin) authenticated as long as there is a `*models.Record` (or `*models.Admin`) instance in the related request context store.

Or in other words, the middleware could look something like:

```go
package main

import (
  "log"

  "github.com/labstack/echo/v5"
  "github.com/pocketbase/pocketbase"
  "github.com/pocketbase/pocketbase/apis"
  "github.com/pocketbase/pocketbase/core"
  "github.com/pocketbase/pocketbase/models"
)

func web3authMiddleware(app core.App) echo.MiddlewareFunc {
  return func(next echo.HandlerFunc) echo.HandlerFunc {
    return func(c echo.Context) error {
      token := c.Request().Header.Get("Authorization")

      // do some token validations (if the token is set)
      // and extract the user identifier from the validated token payload (eg. email)...
      email := ...

      // "sing-in" the found auth record (if any)
      // or maybe create a new auth record if missing (see https://pocketbase.io/docs/record-methods/#create-new-record)
      authRecord, _ := app.Dao().FindAuthRecordByEmail("users", email)
      if authRecord != nil {
        c.Set(apis.ContextAuthRecordKey, authRecord)
      }

      // proceed with the request
      return next(c)
    }
  }
}

func main() {
  app := pocketbase.New()

  app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
    e.Router.Use(web3authMiddleware(app))

    return nil
  })

  if err := app.Start(); err != nil {
    log.Fatal(err)
  }
}
```

_(if you haven't already, you can also check https://pocketbase.io/docs/router for more details)_

**Reply by vpavlin** [reply]
Awesome, actually was able to get exactly here based on the example in the Router Docs section and by modyfing https://github.com/pocketbase/pocketbase/blob/master/apis/middlewares.go#L213 to basically what you posted.

It seems I am getting there. I'll post the final code once done

THanks a lot for quick reply!

**Comment by vpavlin** [comment]
I have a follow up question, if I may.

When I use the external wallet provider with web3auth, JWT only contains the wallet address, but not the email. THat's fine, but I would like to let the user to add email later.

I found the `requestEmailChange` and `confirmEmailChange` but the latter requires password, which the user does not have and if not provided, I get that the password cannot be blank. Not sure if there is a way around it, because aparently auth collections cannot be modified by users.

I could magine adding another route for confirmation, verifying web3auth JWT and the JWT token generated for email change and then updating directly, but wanted to see if that is the way to go

**Reply by ganigeorgiev** [reply]
Hm, this is a use case that could happen also if the user account was created with an OAuth2 provider. Users without an email currently have to set a password first and then request an email change so that they can set their email, which is not very user friendly.

This will need to be improved. Maybe the easiest solution would be to allow users with _empty email_ field to submit their address directly as part of the regular update action (_`verified` will be updated to `false`, until they confirm it with a verification email_) and also the opposite - allow users to clear their email in case the `email` is not marked required in the collection options. Or maybe don't require a password for the confirmEmailChange if the previous user email was empty.
I'll have to think a little more on it and eventually will be resolved in the next v0.14/v0.15 release.

---

> I could magine adding another route for confirmation, verifying web3auth JWT and the JWT token generated for email change and then updating directly, but wanted to see if that is the way to go

Yes, this could be one possible workaround. Alternatively, you can also define a middleware that'll intercept the regular `/collections/users/confirm-email-change` route (eg. via `c.Path()`) by some criteria based on the request parameters.

**Reply by vpavlin** [reply]
I guess the differnce is that with OAuth2 provider you always get the email (I think?) and you are not going to change it (I guess?:D). But maybe there are case where it can happen that the email ends up being empty.

So yes, passwordless change of the email (even if just the empty one) would be great:D

For my case, I might actually go with the custom route and expect the JWT + a signature by the user wallet to confirm the email change to make it more "web3 native".

**Reply by ganigeorgiev** [reply]
> But maybe there are case where it can happen that the email ends up being empty.

Yes, there are some providers that don't return an email (or the user didn't allow sharing it), so the `email` update for this scenario will have to be improved. I'll try to prioritize it after the [ongoing index optimization tasks](https://github.com/orgs/pocketbase/projects/2) for the next release.

**Reply by vpavlin** [reply]
Sounds good! Thanks a lot, I am super excited about PocketBase:) You are doing amazing job!
